File: stage1/week3/assignment.py
import urllib.request as req
import json
import re
import csv

# 取得資料
url = "https://padax.github.io/taipei-day-trip-resources/taipei-attractions-assignment.json"
with req.urlopen(url) as response:
    data = response.read().decode("utf-8")

# 解析資料，將資料從 json 轉為 dict
data_todic = json.loads(data)

result = data_todic["result"]["results"]


# 取出地址的區域
def get_district(address):
    pattern = r"\s([^\s]+區)"
    parse = re.search(pattern, address)

    if parse:
        district = parse.group(1)
    else:
        district = ""

    return district


# 取出第一張圖檔網址
def get_first_image(file):
    file_lower = file.lower()
    image = file_lower.find(".jpg")
    first_image = file[: image + 4]

    return first_image


def attraction_data(attraction):
    attraction_name = data["stitle"]
    attraction_address = data["address"]  # 再處理 取區域
    attraction_longitude = data["longitude"]
    attraction_latitude = data["latitude"]
    attraction_file = data["file"]  # 再處理 取第一筆

    form = [
        attraction_name,
        get_district(attraction_address),
        attraction_longitude,
        attraction_latitude,
        get_first_image(attraction_file),
    ]
    return form


def group_attractions_by_mrt(result):
    attraction_by_mrt = {}

    for data in result:
        attraction_name = data["stitle"]
        attraction_mrt = data["MRT"]

        if attraction_mrt in attraction_by_mrt:
            attraction_by_mrt[attraction_mrt].append(attraction_name)
        else:
            attraction_by_mrt[attraction_mrt] = [attraction_name]

    return attraction_by_mrt


# 寫入 attracion.csv 檔案
with open("attracion.csv", "w", newline="") as file:
    writer = csv.writer(file)

    for data in result:
        form = attraction_data(data)
        writer.writerow(form)

# 寫入 mrt.csv 檔案
with open("mrt.csv", "w", newline="") as file:
    writer = csv.writer(file)
    attraction_by_mrt = group_attractions_by_mrt(result)
    for mrt, attractions in attraction_by_mrt.items():
        form = [mrt] + attractions
        writer.writerow(form)

File: stage1/week3/assignment.js
// 建立照片模板
function createAndAppendElementPic(tag, className, src, container) {
    const element = document.createElement(tag);
    element.classList.add(className);    
    const img = document.createElement("img");
    img.src = src;
    element.appendChild(img);
    
    container.appendChild(element)
}

// 建立景點名稱模板
function createAndAppendElementText(tag, className, title, container) {
    const element = document.createElement(tag);
    element.classList.add(className);
    element.textContent = title;
    container.appendChild(element)
}

// 處理取得的資料並套入模板
function processDataAndDisplay(data, isProduct){
    data.forEach(function(item) {
        const {file, stitle} = item;
        // 取得圖片的第一個網址
        let file_lower = file.toLowerCase();
        const image = file_lower.indexOf(".jpg");
        const first_image = file.substring(0, image + 4);

        // 建立最外層區塊
        if (isProduct) {
            const container = document.getElementById("product-container")
            const productItem = document.createElement("div");
            productItem.classList.add("product-item");

            createAndAppendElementPic("div", "attractionPic", first_image, productItem)
            createAndAppendElementText("div", "attractionName", stitle, productItem)

            container.appendChild(productItem)
        }
        else {
            const container = document.getElementById("attractions-container")
            const titleItem = document.createElement("div");
            titleItem.classList.add("title-item");

            createAndAppendElementPic("div", "attractionPic", first_image, titleItem)
            createAndAppendElementPic("div", "star", "./star.png", titleItem)
            createAndAppendElementText("div", "attractionName", stitle, titleItem)
            

            container.appendChild(titleItem);
        }
        
    })

}


// 爬取資料
fetch("https://padax.github.io/taipei-day-trip-resources/taipei-attractions-assignment.json")
    .then(response => response.json())
    .then(function(responseData) {
        const productData = responseData.result.results.slice(0,3);
        const articleData = responseData.result.results.slice(3,15);
        processDataAndDisplay(productData, true);
        processDataAndDisplay(articleData, false);
    })
    

// 設定load more按鈕
let currentIndex = 15;
const totalAttractions = 58;
const loadMoreButton = document.getElementById("load-more");
loadMoreButton.addEventListener("click", function() {
    fetch("https://padax.github.io/taipei-day-trip-resources/taipei-attractions-assignment.json")
    .then(response => response.json())
    .then(function(responseData) {
        data = responseData.result.results;
        const endIndex = Math.min(currentIndex + 12, totalAttractions);
        articleData = data.slice(currentIndex, endIndex);
    
        if (articleData.length >= 12) {
            processDataAndDisplay(articleData, false);
            currentIndex = endIndex;

            if (currentIndex === totalAttractions) {
                loadMoreButton.disabled = true;
            }
        }
        else {
            const articleData = data.slice(currentIndex);
            processDataAndDisplay(articleData, false);
            loadMoreButton.disabled = true;
        }
    })
})
